hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/TestDFSMkdirs.java

```diff
@@ -152,4 +154,80 @@ public void testMkdirRpcNonCanonicalPath() throws IOException {
       cluster.shutdown();
     }
   }
+
+  @Test
+  public void testMkDirsWithRestart() throws IOException {
+    MiniDFSCluster cluster =
+        new MiniDFSCluster.Builder(conf).numDataNodes(2).build();
+    DistributedFileSystem dfs = cluster.getFileSystem();
+    try {
+      // Create a dir in root dir, should succeed
+      assertTrue(dfs.mkdir(new Path("/mkdir-1"), FsPermission.getDefault()));
+      dfs.mkdir(new Path("/mkdir-2"), FsPermission.getDefault());
+      dfs.mkdir(new Path("/mkdir-3"), FsPermission.getDefault());
+      DFSTestUtil.writeFile(dfs, new Path("/mkdir-1/file1"), "hello world");
+      cluster.restartNameNodes();
+      dfs = cluster.getFileSystem();
+      assertTrue(dfs.exists(new Path("/mkdir-1")));
+      assertTrue(dfs.exists(new Path("/mkdir-2")));
+      assertTrue(dfs.exists(new Path("/mkdir-3")));
+      assertTrue(dfs.exists(new Path("/mkdir-1/file1")));
+    } finally {
+      dfs.close();
+      cluster.shutdown();
+    }
+  }
+
+  @Test
+  public void testMkdirWithDelete() throws IOException {
```

Comment:
Can you combine these 2 tests in one. First assert the delete logic, then restart NN and check the existence of `mkdir-?`
The problem with TestDFSMkdirs that it creates MiniDFSCluter for each test case. We don't want to refactor it on the branch, but would prefer to minimize replicating the wrong pattern.